In [2]:
import tensorflow as tf
import numpy as np
import sys
sys.path.append("..") # Adds higher directory to python modules path.
import branchingdnn as branching
from branchingdnn.utils import * 

In [3]:
normal_dataset = branching.dataset.prepare.dataset(tf.keras.datasets.cifar10.load_data(),32,5000,22500,(227,227))

trainSize 45000
testSize 10000


In [29]:
x = tf.keras.models.load_model('../models/alexNetv6_vanilla_T_2.hdf5', custom_objects={'confidenceScore': confidenceScore,
                                                                         'unconfidence': unconfidence,
                                                                         'confidenceDifference': confidenceDifference,
                                                                         'BranchEndpoint': branching.branches.branch.BranchEndpoint,
                                                                         'FeatureDistillation': branching.branches.branch.FeatureDistillation})

In [30]:
x.compile(optimizer=tf.optimizers.SGD(lr=0.001), loss='SparseCategoricalCrossentropy', metrics=['accuracy',confidenceDifference],run_eagerly=True)
train_ds, test_ds, validation_ds = normal_dataset
x.evaluate(test_ds, verbose=2)

312/312 - 19s - loss: 4.0657 - dense_2_loss: 0.6242 - branch_softmax_loss: 1.5308 - branch_softmax_1_loss: 1.1518 - branch_softmax_2_loss: 0.7589 - dense_2_accuracy: 0.7948 - dense_2_confidenceDifference: 0.0955 - branch_softmax_accuracy: 0.6240 - branch_softmax_confidenceDifference: 0.0548 - branch_softmax_1_accuracy: 0.7307 - branch_softmax_1_confidenceDifference: 0.0603 - branch_softmax_2_accuracy: 0.8082 - branch_softmax_2_confidenceDifference: 0.0702


[4.065739631652832,
 0.6242280006408691,
 1.5308417081832886,
 1.1518110036849976,
 0.7588583827018738,
 0.7947716116905212,
 0.09554122388362885,
 0.6239984035491943,
 0.05482248589396477,
 0.7306690812110901,
 0.06027385964989662,
 0.8081930875778198,
 0.07017481327056885]

In [31]:
x = tf.keras.models.load_model('../models/alexNetv6_feat_distill_3.hdf5', custom_objects={'confidenceScore': confidenceScore,
                                                                         'unconfidence': unconfidence,
                                                                         'confidenceDifference': confidenceDifference,
                                                                         'BranchEndpoint': branching.branches.branch.BranchEndpoint,
                                                                         'FeatureDistillation': branching.branches.branch.FeatureDistillation,
                                                                         'FeatureDistillation_clear': branching.branches.branch.FeatureDistillation_clear})

In [32]:
x.summary()

Model: "model_branched"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 227, 227, 3) 0                                            
__________________________________________________________________________________________________
conv2d (Conv2D)                 (None, 55, 55, 96)   34944       input_1[0][0]                    
__________________________________________________________________________________________________
batch_normalization (BatchNorma (None, 55, 55, 96)   384         conv2d[0][0]                     
__________________________________________________________________________________________________
max_pooling2d (MaxPooling2D)    (None, 27, 27, 96)   0           batch_normalization[0][0]        
_____________________________________________________________________________________

In [20]:
if not dataset:
    dataset = branching.dataset.prepare.dataset_distil(tf.keras.datasets.cifar10.load_data(),32,5000,22500,(227,227))

KeyboardInterrupt: 

In [27]:
x.compile(optimizer=tf.optimizers.SGD(lr=0.001), loss='SparseCategoricalCrossentropy', metrics=['accuracy',confidenceDifference],run_eagerly=True)

In [28]:
train_ds, test_ds, validation_ds = dataset
x.evaluate(test_ds, verbose=2)

312/312 - 19s - loss: 106350.4688 - dense_2_loss: 0.6242 - branch_softmax_loss: 1.0999 - dense_2_accuracy: 0.7948 - dense_2_confidenceDifference: 0.0955 - branch_softmax_accuracy: 0.6580 - branch_softmax_confidenceDifference: 0.0683 - branch_teaching_distill: 106347.7500 - branch_softmax_KL: 0.9727


[106350.46875,
 0.6242280006408691,
 1.0999187231063843,
 0.7947716116905212,
 0.09554122388362885,
 0.6579527258872986,
 0.06826221942901611,
 106347.75,
 0.9727215766906738]

baseline: 
312/312 - 19s - loss: 3.6760 - dense_2_loss: 0.6242 - branch_softmax_loss: 1.3435 - branch_softmax_1_loss: 1.0038 - branch_softmax_2_loss: 0.7045 - dense_2_accuracy: 0.7948 - dense_2_confidenceDifference: 0.0955 - branch_softmax_accuracy: 0.6202 - branch_softmax_confidenceDifference: 0.0610 - branch_softmax_1_accuracy: 0.7387 - branch_softmax_1_confidenceDifference: 0.0632 - branch_softmax_2_accuracy: 0.8073 - branch_softmax_2_confidenceDifference: 0.0773

feature_distil 1 results:

312/312 - 19s - loss: 106350.4688 - dense_2_loss: 0.6242 - branch_softmax_loss: 1.0999 - dense_2_accuracy: 0.7948 - dense_2_confidenceDifference: 0.0955 - branch_softmax_accuracy: 0.6580 - branch_softmax_confidenceDifference: 0.0683 - branch_teaching_distill: 106347.7500 - branch_softmax_KL: 0.9727

number 2 results:
312/312 - 19s - loss: 106322.7031 - dense_2_loss: 0.6242 - branch_softmax_loss: 1.2637 - dense_2_accuracy: 0.7948 - dense_2_confidenceDifference: 0.0955 - branch_softmax_accuracy: 0.6382 - branch_softmax_confidenceDifference: 0.0631 - branch_teaching_distill: 106319.7344 - branch_softmax_KL: 1.0670


number 3 results:
312/312 - 29s - loss: 166223.5781 - dense_2_loss: 0.6242 - branch_softmax_loss: 1.1222 - branch_softmax_1_loss: 0.8409 - branch_softmax_2_loss: 0.6892 - dense_2_accuracy: 0.7948 - dense_2_confidenceDifference: 0.0955 - branch_softmax_accuracy: 0.6825 - branch_softmax_confidenceDifference: 0.0667 - branch_softmax_1_accuracy: 0.7635 - branch_softmax_1_confidenceDifference: 0.0700 - branch_softmax_2_accuracy: 0.8064 - branch_softmax_2_confidenceDifference: 0.0775 - branch_teaching_distill: 104909.4453 - branch_teaching_1_distill: 61309.5430 - branch_softmax_KL: 0.8163 - branch_softmax_1_KL: 0.3491 - branch_softmax_2_KL: 0.0748

In [16]:
num = 0.6825 - 0.6202 
num

0.024499999999999966

In [43]:
branches = 1
teaching_features = [1,2,3,4]
if type(teaching_features) is not list:
    teaching_features = [teaching_features]
print(teaching_features)
print(min(branches, len(teaching_features)))
y = teaching_features[min(branches, len(teaching_features))]
y

[1, 2, 3, 4]
1


2